src/solvers/refinement/axt.h

```diff
@@ -0,0 +1,55 @@
+/*******************************************************************\
+
+ Module: Expression Building Utility
+
+ Author: Diffblue Limited. All rights reserved.
+
+\*******************************************************************/
+#ifndef CPROVER_SOLVERS_REFINEMENT_AXT_H
+#define CPROVER_SOLVERS_REFINEMENT_AXT_H
+
+#include <util/expr.h>
+#include <util/string_expr.h>
+
+/// Thin wrapper for exprt class assigning different meaning to operators
+/// Performing logical expression on axt will not yield usual results
+/// results (with exception of assignment operator), but rather new axt
+/// instances containing trees of logical expressions expressed as exprt
+/// hierarchies
+class axt final
+{
+public:
+  axt(const exprt& expr) // NOLINT Use implicit conversions for best results
+    :expr_(expr) { }
+  axt operator>(const axt& rhs) const
+  { return axt(binary_relation_exprt(expr_, ID_gt, rhs.expr_)); }
+  axt operator<(const axt& rhs) const
+  { return axt(binary_relation_exprt(expr_, ID_lt, rhs.expr_)); }
+  axt operator>=(const axt& rhs) const
+  { return !(*this<rhs); }
+  axt operator<=(const axt& rhs) const
+  { return !(*this>rhs); }
+  axt operator==(const axt& rhs) const
+    { return axt(equal_exprt(expr_, rhs.expr_)); }
+  // Implication
+  axt operator>>=(const axt& rhs) const
+  { return axt(implies_exprt(expr_, rhs.expr_)); }
+  axt operator+(const axt& rhs) const
+  { return axt(plus_exprt(expr_, rhs.expr_)); }
+  axt operator-(const axt& rhs) const
+  { return axt(minus_exprt(expr_, rhs.expr_)); }
+  axt operator&&(const axt& rhs) const
+  { return axt(and_exprt(expr_, rhs.expr_)); }
+  axt operator||(const axt& rhs) const
+  { return axt(or_exprt(expr_, rhs.expr_)); }
+  axt operator[](const axt& index) const
+  { return axt(to_string_expr(expr_)[index.expr_]); }
```

Comment:
Why do we cast to a string expression? Is this only going to be for the string solver?

Reply:
For now it's only used in the string solver, but ultimately this should represent `index_of_exprt` (taking an `array_exprt` on the left) this will be unified once the new version of the string solver is merged as strings will be represented by arrays.